[RFC] main: Add a warning if runtimepath doesn't point to a Neovim runtime

Zashas opened the pull request:
This commit let nvim show a warning if $VIMRUNTIME doesn't point to a valid Neovim runtime : 

```
[math@seattle neovim]$ export VIMRUNTIME=/usr/share/vim/vim74/
[math@seattle neovim]$ ./build/bin/nvim 
EXX: Invalid Neovim runtime from $VIMRUNTIME=/usr/share/vim/vim74/
Press ENTER or type command to continue
```

Resolves issue #2977. Just took the code of equalsraf (https://github.com/neovim/neovim/pull/1927#issuecomment-96108014) and put it in a function, so all credit goes entirely to him. The function is called before the first use of $VIMRUNTIME (by `init_locale`).

Last thing to do is picking a number error. Can I pick one myself or can someone give me one ?

Zashas committed:
main: Add a warning if runtimepath doesn't point to a Neovim runtime

Pyrohh commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,20 @@ static void check_swap_exists_action(void)
+static void check_vimruntime(void)
+{
+  const char *runtime = os_getenv("VIMRUNTIME");
+  if (runtime && runtime[0]) {
```
You only need to check for nullity. Here's the os_getenv() source:
```c
/// Like getenv(), but returns NULL if the variable is empty.
const char *os_getenv(const char *name)
  FUNC_ATTR_NONNULL_ALL
{
  const char *e = getenv(name);
  return e == NULL || *e == NUL ? NULL : e;
}
```

ZyX-I commented:
> EXX: Invalid Neovim runtime from $VIMRUNTIME=/usr/share/vim/vim74/

I do not like such messages. It should explain why you don’t think that $VIMRUNTIME is not valid.

Also how about testing for some *meaningful* file like `runtime/autoload/provider/clipboard.vim`? E.g. Gentoo in vim-core ebuild with USE=minimal does not install doc file and NeoVim works completely fine without documentation.

justinmk commented:
Neovim works fine without runtime at all. It should be a warning only, and should not cause press-enter prompt.

ZyX-I commented:
Warnings start with `WXX`, not `EXXX`. Almost never saw this feature, but some commands trigger them. I found nine warnings in the codebase, use `ag --ignore po --ignore testdir 'W\d+' src/nvim` to search for them.

Zashas commented:
Commit updated according to your comments. Had to use mch_errmsg for printing the warning, EMSG and MSG mandatory start a press-enter prompt.

I took the next unused warning number, which is 19.

justinmk commented:
Looks good, thanks. Author should be `Rui Abreu Ferreira <raf-ep@gmx.com>`, then bottom of the commit message:

```
Helped-by: Michael Reed <m.reed@mykolab.com>
Helped-by: Mathieu Xhonneux <m.xhonneux@gmail.com>
```

> I took the next unused warning number, which is 19.

The error numbers tend to skip ranges to allow space for error "groups". In our case, it also helps future-compat with upstream Vim. So I suggest `W50`.

justinmk commented:
Also, why do this in `early_init()`? Perhaps a better time would be after `source_startup_scripts()`, before `load_plugins()`, since vimrc may modify `'rtp'`?

justinmk commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,29 @@ static void check_swap_exists_action(void)
+static void check_vimruntime(void)
+{
+  const char *runtime = os_getenv("VIMRUNTIME");
+  if (runtime) {
```
This doesn't help the case where VIMRUNTIME env var is not defined, but runtime is invalid. E.g., if user forgot to install, or does not have permissions.

On one of my machines, I use:

    make CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX:PATH=~/bin/neovim" && make install

If `~/bin/neovim` gets deleted or something, the warning won't appear, but the user will still have the same questions that we are trying to address.

justinmk commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,29 @@ static void check_swap_exists_action(void)
+    append_path(rt_testpath, "provider", MAXPATHL);
+    append_path(rt_testpath, "clipboard.vim", MAXPATHL);
+
+    if (!os_file_exists((char_u *)rt_testpath)) {
```
Can you test what this returns if `rt_testpath` is not readable (e.g. user lacks permissions).

justinmk commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,29 @@ static void check_swap_exists_action(void)
+      char warning_message[MAXPATHL*2];
+      vim_snprintf(warning_message, sizeof(warning_message),
+          _("W19: The VIMRUNTIME specified (%s) is not a valid Neovim runtime."
+            " %s could not be found.\n"),
```
W50: Invalid VIMRUNTIME: (%s)

`:help W50` can simply document the file that is expected to exist. It doesn't need to be in the warning message.

ZyX-I replied:
This is enough, but readily available information is better.

By the way, second %s may be just autoload/provider/clipboard.vim. No need to repeat full runtime.

justinmk commented:
> I do not like such messages. It should explain why you don’t think that $VIMRUNTIME is not valid.

@ZyX-I If `:help W50` explains exactly what defines a valid runtime, isn't that enough?

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,29 @@ static void check_swap_exists_action(void)
+    append_path(rt_testpath, "clipboard.vim", MAXPATHL);
+
+    if (!os_file_exists((char_u *)rt_testpath)) {
+      char warning_message[MAXPATHL*2];
```
I would use IObuff in this case. It is very commonly used for various messages.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,29 @@ static void check_swap_exists_action(void)
+{
+  const char *runtime = os_getenv("VIMRUNTIME");
+  if (runtime) {
+    char rt_testpath[MAXPATHL];
```
And for paths there is NameBuff global.

Zashas commented:
New commit :
- Changed the location of the call to `check_vimruntime` as justinmk has suggested
- I don't rely directly on $VIMRUNTIME anymore. I use the `do_in_runtimepath` function, which checks every possible location of the runtime. 

Apparently, Vim overwrites the $VIMRUNTIME if it hasn't been set by the user to the first runtime he has found (with `rtp`'s value), but I don't know if it's always true, so I added an extra warning. I can dig into that later.

I'm waiting on justinmk's PR #3166 to check if the runtime is readable.

By the way, the builds do print : "W51: Invalid VIMRUNTIME: "/home/travis/nvim-install/share/nvim"", which is, I think, incorrect since it must be /home/travis/nvim-install/share/nvim/runtime. Or I am wrong ?

- [math@seattle neovim]$ VIMRUNTIME=/usr/share/nvim/ ./build/bin/nvim -------
Is not OK, triggers W51 and E484: Can't open file /usr/share/nvim/syntax/syntax.vim
- [math@seattle neovim]$ VIMRUNTIME=/usr/share/nvim/runtime/ ./build/bin/nvim -------
Is OK
- [math@seattle neovim]$ VIM=/usr/share/nvim/ ./build/bin/nvim -------
Is OK

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,37 @@ static void check_swap_exists_action(void)
+    if (runtime) {
+      vim_snprintf((char*) IObuff, IOSIZE,
+        _("W51: Invalid VIMRUNTIME: \"%s\"\n"), runtime);
+    }
```
Closing figure brace must be on the same line as `else`.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,37 @@ static void check_swap_exists_action(void)
+    const char *runtime = os_getenv("VIMRUNTIME");
+    if (runtime) {
+      vim_snprintf((char*) IObuff, IOSIZE,
+        _("W51: Invalid VIMRUNTIME: \"%s\"\n"), runtime);
```
Must be aligned with the first argument. You should either move first two arguments to the next line or align `_` with `(` in `(char*)`.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,37 @@ static void check_swap_exists_action(void)
+static void check_vimruntime(void)
+{
+  int result = do_in_runtimepath((char_u*) "autoload/provider/clipboard.vim",
+    FALSE, check_vimruntime_callback, NULL);
```
`FALSE` must not be used in the new code. Use `false` from `stdbool.h`.

ZyX-I replied:
This code is also misaligned like `vim_snprintf` calls.

justinmk replied:
Off-topic, I just noticed that `do_in_runtimepath(NULL, TRUE, foo, NULL)` will cause a null deference.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,37 @@ static void check_swap_exists_action(void)
+    }
+    else {
+      vim_snprintf((char*) IObuff, IOSIZE,
+       _("W50: Neovim's runtime couldn't be found.\n"));
```
Misaligned code. In addition to the comment to line 2099 `_` is also 1-shifted.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,37 @@ static void check_swap_exists_action(void)
+        _("W51: Invalid VIMRUNTIME: \"%s\"\n"), runtime);
+    }
+    else {
+      vim_snprintf((char*) IObuff, IOSIZE,
```
You don’t need `vim_snprintf` here. Just use `mch_errmsg(_("W50: …"))` directly.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,37 @@ static void check_swap_exists_action(void)
 }
+
+
+// Used by check_vimruntime
```
Should be doxygen-style comment.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,37 @@ static void check_swap_exists_action(void)
+  STRNCPY(NameBuff, fname, MAXPATHL);
+}
+
+// Check if $VIMRUNTIME points to a valid Neovim runtime path,
```
Doxygen-style.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2072,37 @@ static void check_swap_exists_action(void)
+
+  if (result == OK) {
+      //check if file in NameBuff can be read, W52 if not
+  }
```
Line must be joined with the next one.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2073,35 @@ static void check_swap_exists_action(void)
+  STRNCPY(NameBuff, fname, MAXPATHL);
+}
+
+/// Check if $VIMRUNTIME points to a valid Neovim runtime path, look for
```
Current version checks not $VIMRUNTIME.

ZyX-I commented on src/nvim/main.c:
```diff
@@ -2070,3 +2073,35 @@ static void check_swap_exists_action(void)
+    const char *runtime = os_getenv("VIMRUNTIME");
+    if (runtime) {
+      vim_snprintf((char*) IObuff, IOSIZE,
+                   _("W51: Invalid VIMRUNTIME: \"%s\""), runtime);
```
This error message is invalid. The fact that VIMRUNTIME has some non-empty value does *not* mean that it is present in runtimepath.

ZyX-I commented:
By the way, I do not want to see *any* warnings when testing my plugins. And I am not going to use NeoVim runtime there. When it was checking $VIMRUNTIME that was standable: I could simply exclude it from the &runtimepath while keeping valid runtime there. When searching inside &runtimepath it is not.

Zashas commented:
Well, that's what @justinmk was looking for : https://github.com/neovim/neovim/pull/3159#discussion_r36932567

What we could do instead is checking for a Vim runtime-file, and then for a Neovim specific-file. If the latter isn't present and the former is, then Neovim is using Vim's runtime, and that could trigger the warning. This warning would not be triggered if Nvim doesn't find any runtime, since as justinmk said, Nvim can work fine without the runtime.

justinmk commented:
@ZyX-I your complaint is not able to be satisfied because there are plans for other warnings besides this. And the messages maintainer line is going to be removed...

justinmk commented:
@Zashas https://github.com/neovim/neovim/pull/3166 is merged now.

Zashas commented:
I finalized the commit. It works fine on my setup and passes all tests, yet the unittests do fail on Travis. There's no proper error message, so I don't know what went wrong. Any clue ?

Zashas commented:
I added the warning W51 in the docs.

I think that everything is OK. I made some tests, and Travis fails because the warning 51 is triggered : the runtime is not found, I guess it's not installed or not properly indicated on Travis.

I'm changing this PR to RFC.

fmoralesc commented:
I was just taking a look at this, and thought: shouldn't the warning be given only if the clipboard file is not found, *but* some other file supposed to exist in both vim and neovim's runtime *does* exist? Basically, the issue is the scenario when the vim runtime is used instead of ours. Otherwise running nvim without the runtime will always cause the warning, and this is a problem for people who don't want to use the default runtime and for the use-case of the Travis builds, where apparently it's not available... What do you think, @justinmk and @Zashas?

fwalch commented:
@fmoralesc Or, just warn if a file that's only in Vim exists (e.g. `doc/version7.txt`)?

fmoralesc commented:
@fwalch That would do it too :+1:

justinmk commented:
@fmoralesc @fwalch :+1: `doc/version7.txt` is a good idea, let's go with that.

In the case where `provider/clipboard.vim` is not found *and* `doc/version7.txt` is not found, there should be a different (silent) hint in `:messages`. That can be a different PR though, it needs https://github.com/neovim/neovim/pull/1802

dsummersl commented:
Anything I can do to help this along - its marked for the .2 milestone...

justinmk commented:
@dsummersl It's blocked by #1802.

dsummersl commented:
Ah, k - so thats not in this milestone. Should #1802 go into 0.2, or should #3159 get booted out?

justinmk commented:
Fixed :sweat_smile:

dsummersl commented:
:+1:
